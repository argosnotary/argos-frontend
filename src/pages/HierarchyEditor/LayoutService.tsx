/*
 * Copyright (C) 2020 Argos Notary
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  ILayout,
  ILayoutMetaBlock,
  ILayoutSegment,
  IRule,
  IMatchRule,
  RuleRuleTypeEnum,
  IStep
} from "../../interfaces/ILayout";
import stringify from "json-stable-stringify";
import { signString } from "../../security";

const signLayout = async (
  password: string,
  keyId: string,
  encryptedPrivateKey: string,
  layout: ILayout
): Promise<ILayoutMetaBlock> => {
  const signature = await signString(
    password,
    encryptedPrivateKey,
    serialize(layout)
  );
  return {
    signatures: [{ signature: signature, keyId: keyId, keyAlgorithm: "EC", hashAlgorithm: "SHA384" }],
    layout: layout
  };
};

const serialize = (layout: ILayout): string => {
  return stringify(mapLayout(layout));
};

const mapLayout = (layout: ILayout): ILayout => {
  return {
    keys: layout.keys.map(key => {
      return { ...key };
    }),
    authorizedKeyIds: layout.authorizedKeyIds,
    expectedEndProducts: layout.expectedEndProducts.map(rule => {
      return { ...rule, ruleType: RuleRuleTypeEnum.MATCH };
    }),
    layoutSegments: layout.layoutSegments.map(mapSegment).sort((n1, n2) => {
      if (n1.name > n2.name) {
        return 1;
      }

      if (n1.name < n2.name) {
        return -1;
      }

      return 0;
    })
  };
};

const mapSegment = (segment: ILayoutSegment): ILayoutSegment => {
  return {
    name: segment.name,
    steps: segment.steps.map(mapStep).sort((n1, n2) => {
      if (n1.name > n2.name) {
        return 1;
      }

      if (n1.name < n2.name) {
        return -1;
      }

      return 0;
    })
  };
};

const mapStep = (step: IStep): IStep => {
  return {
    ...step,
    expectedMaterials: step.expectedMaterials.map(mapRule),
    expectedProducts: step.expectedProducts.map(mapRule)
  };
};

const mapRule = (rule: IRule): IRule | IMatchRule => {
  if (rule.ruleType === RuleRuleTypeEnum.MATCH) {
    return {
      ...rule,
      ruleType: RuleRuleTypeEnum.MATCH
    };
  } else {
    return { ruleType: rule.ruleType || RuleRuleTypeEnum.ALLOW, pattern: rule.pattern };
  }
};

export { serialize, signLayout };
