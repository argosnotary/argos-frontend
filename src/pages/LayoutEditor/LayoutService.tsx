/*
 * Copyright (C) 2019 - 2020 Rabobank Nederland
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  ILayout,
  ILayoutMetaBlock,
  ILayoutSegment,
  IRule,
  RuleRuleTypeEnum,
  IStep
} from "../../interfaces/ILayout";
import { Domain } from "./argos-domain.layout";
import stringify from "json-stable-stringify";
import { signString } from "../../security";

const signLayout = async (
  password: string,
  keyId: string,
  encryptedPrivateKey: string,
  layout: ILayout
): Promise<ILayoutMetaBlock> => {
  const signature = await signString(
    password,
    encryptedPrivateKey,
    serialize(layout)
  );
  return {
    signatures: [{ signature: signature, keyId: keyId }],
    layout: layout
  };
};

const serialize = (layout: ILayout): string => {
  return stringify(mapLayout(layout));
};

const mapLayout = (layout: ILayout): Domain.ILayout => {
  return {
    keys: layout.keys.map(key => {
      return { ...key };
    }),
    authorizedKeyIds: layout.authorizedKeyIds,
    expectedEndProducts: layout.expectedEndProducts.map(rule => {
      return { ...rule, ruleType: "MATCH" };
    }),
    layoutSegments: layout.layoutSegments.map(mapSegment).sort((n1, n2) => {
      if (n1.name > n2.name) {
        return 1;
      }

      if (n1.name < n2.name) {
        return -1;
      }

      return 0;
    })
  };
};

const mapSegment = (segment: ILayoutSegment): Domain.ILayoutSegment => {
  return {
    name: segment.name,
    steps: segment.steps.map(mapStep).sort((n1, n2) => {
      if (n1.name > n2.name) {
        return 1;
      }

      if (n1.name < n2.name) {
        return -1;
      }

      return 0;
    })
  };
};

const mapStep = (step: IStep): Domain.IStep => {
  return {
    ...step,
    expectedMaterials: step.expectedMaterials.map(mapRule),
    expectedProducts: step.expectedProducts.map(mapRule)
  };
};

const mapRule = (rule: IRule): Domain.IRule | Domain.IMatchRule => {
  if (rule.ruleType === RuleRuleTypeEnum.MATCH) {
    return {
      ...rule,
      ruleType: "MATCH"
    };
  } else {
    return { ruleType: rule.ruleType, pattern: rule.pattern };
  }
};

export { serialize, signLayout };
