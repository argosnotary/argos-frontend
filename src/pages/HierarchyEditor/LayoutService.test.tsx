/*
 * Copyright (C) 2019 - 2020 Rabobank Nederland
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { serialize } from "./LayoutService";
import {
  RuleDestinationTypeEnum,
  RuleRuleTypeEnum
} from "../../interfaces/ILayout";

const PUBLIC_KEY =
  "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAoDIw5p3LYjFLr+JFvWOz0KQ22x538O2BoQO4e4EorYYkSMHn00pabqAQ9z6+nA+l43+Jxb5eJrHoroV4YZN8WDMsjY0eB1Q1K6hl3SHgeqTvA2i2GlpDg7zLRnP9YWbUwbWP+UWtRFK0x1lCCkSmxsP2HAom/T11/MMd/kitVt0rGsq8wQH7PLsOZ8zPh4sQ0iyCVLil6+VF6zsT83dKFocdfZWAywkQ6sVZbuzFCe+pLQktwTz1Ir8mMQi6sPh57b5yyFCSVstK1lKf+OQTtuuQzYz2bvpr9zkXr0O80IdTXOnoO1vM1lJuRPT3J0Zcr2nYdbmIskp4ZQyezXMqawIDAQAB";

it("test serialize", () => {
  const seralized = serialize({
    keys: [{ id: "keyId", key: PUBLIC_KEY }],
    expectedEndProducts: [
      {
        destinationSegmentName: "destinationSegmentName",
        destinationType: RuleDestinationTypeEnum.PRODUCTS,
        destinationStepName: "destinationStepName",
        pattern: "MatchFiler",
        destinationPathPrefix: "destinationPathPrefix",
        sourcePathPrefix: "sourcePathPrefix"
      }
    ],
    layoutSegments: [
      {
        name: "segment 1",
        steps: [
          {
            name: "step b",
            requiredNumberOfLinks: 1,
            expectedMaterials: [
              { ruleType: RuleRuleTypeEnum.ALLOW, pattern: "AllowRule" },
              {
                ruleType: RuleRuleTypeEnum.REQUIRE,
                pattern: "RequireRule"
              }
            ],
            expectedProducts: [
              { ruleType: RuleRuleTypeEnum.CREATE, pattern: "CreateRule" },
              {
                ruleType: RuleRuleTypeEnum.MODIFY,
                pattern: "ModifyRule"
              }
            ],
            authorizedKeyIds: ["key2", "key1"]
          },
          {
            name: "step a",
            authorizedKeyIds: ["step a key 2", "step a key 1"],
            requiredNumberOfLinks: 23,
            expectedProducts: [
              {
                ruleType: RuleRuleTypeEnum.DISALLOW,
                pattern: "DisAllowRule"
              },
              {
                destinationPathPrefix: "destinationPathPrefix",
                sourcePathPrefix: "sourcePathPrefix",
                destinationStepName: "destinationStepName",
                destinationSegmentName: "segment 1",
                destinationType: RuleDestinationTypeEnum.MATERIALS,
                pattern: "MatchRule",
                ruleType: RuleRuleTypeEnum.MATCH
              }
            ],
            expectedMaterials: []
          }
        ]
      }
    ],
    authorizedKeyIds: ["key2", "key1"]
  });

  expect(seralized).toMatchSnapshot();
});
