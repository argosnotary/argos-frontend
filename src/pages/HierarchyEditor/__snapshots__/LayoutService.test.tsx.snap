// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test serialize 1`] = `"{\\"authorizedKeyIds\\":[\\"key2\\",\\"key1\\"],\\"expectedEndProducts\\":[{\\"destinationPathPrefix\\":\\"destinationPathPrefix\\",\\"destinationSegmentName\\":\\"destinationSegmentName\\",\\"destinationStepName\\":\\"destinationStepName\\",\\"destinationType\\":\\"PRODUCTS\\",\\"pattern\\":\\"MatchFiler\\",\\"ruleType\\":\\"MATCH\\",\\"sourcePathPrefix\\":\\"sourcePathPrefix\\"}],\\"keys\\":[{\\"algorithm\\":\\"EC\\",\\"keyId\\":\\"keyId\\",\\"publicKey\\":\\"MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAoDIw5p3LYjFLr+JFvWOz0KQ22x538O2BoQO4e4EorYYkSMHn00pabqAQ9z6+nA+l43+Jxb5eJrHoroV4YZN8WDMsjY0eB1Q1K6hl3SHgeqTvA2i2GlpDg7zLRnP9YWbUwbWP+UWtRFK0x1lCCkSmxsP2HAom/T11/MMd/kitVt0rGsq8wQH7PLsOZ8zPh4sQ0iyCVLil6+VF6zsT83dKFocdfZWAywkQ6sVZbuzFCe+pLQktwTz1Ir8mMQi6sPh57b5yyFCSVstK1lKf+OQTtuuQzYz2bvpr9zkXr0O80IdTXOnoO1vM1lJuRPT3J0Zcr2nYdbmIskp4ZQyezXMqawIDAQAB\\"}],\\"layoutSegments\\":[{\\"name\\":\\"segment 1\\",\\"steps\\":[{\\"authorizedKeyIds\\":[\\"step a key 2\\",\\"step a key 1\\"],\\"expectedMaterials\\":[],\\"expectedProducts\\":[{\\"pattern\\":\\"DisAllowRule\\",\\"ruleType\\":\\"DISALLOW\\"},{\\"destinationPathPrefix\\":\\"destinationPathPrefix\\",\\"destinationSegmentName\\":\\"segment 1\\",\\"destinationStepName\\":\\"destinationStepName\\",\\"destinationType\\":\\"MATERIALS\\",\\"pattern\\":\\"MatchRule\\",\\"ruleType\\":\\"MATCH\\",\\"sourcePathPrefix\\":\\"sourcePathPrefix\\"}],\\"name\\":\\"step a\\",\\"requiredNumberOfLinks\\":23},{\\"authorizedKeyIds\\":[\\"key2\\",\\"key1\\"],\\"expectedMaterials\\":[{\\"pattern\\":\\"AllowRule\\",\\"ruleType\\":\\"ALLOW\\"},{\\"pattern\\":\\"RequireRule\\",\\"ruleType\\":\\"REQUIRE\\"}],\\"expectedProducts\\":[{\\"pattern\\":\\"CreateRule\\",\\"ruleType\\":\\"CREATE\\"},{\\"pattern\\":\\"ModifyRule\\",\\"ruleType\\":\\"MODIFY\\"}],\\"name\\":\\"step b\\",\\"requiredNumberOfLinks\\":1}]}]}"`;
